man/mhmail.man
.TH MHMAIL %manext1% 2012-07-01 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
mhmail \- send or read mail non-interactively with nmh
.SH SYNOPSIS
.HP 5
.na
.B mhmail
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-to ]
.I addrs
\&...
.RB [ \-attach
.IR file ]
.RB [ \-body
.IR text ]
.RB [ \-cc
.I addrs
\&...]
.RB [ \-from
.IR addr ]
.RB [ \-headerfield
.IR name:value ]
.RB [ \-subject
.IR subject ]
.RB [ \-resent ]
[switches\ for
.I post
\&... |
.B \-profile
[switches\ for
.I send
\&...]]
.RB [ \-send " | " \-nosend ]
.HP 5
.B mhmail
with no arguments is equivalent to
.B inc
.ad
.SH DESCRIPTION
.B mhmail
is intended as a replacement, compatible with
.BR nmh ,
for the standard Berkeley mail program
.RB ( mail
or
.BR mailx ).
.B nmh
is intended for the use of programs such as
.BR cron ,
which expect to send mail automatically.  It is also used by some
.B nmh
commands to mail error notifications.  Although
.B mhmail
can be used interactively, it is recommended that
.B comp
and
.B send
be used instead, to send messages.
.PP
When invoked without arguments,
.B mhmail
simply invokes
.B inc
to incorporate new messages from the user's mail drop.
.PP
When one or more addresses are specified, a message is read from the
standard input and spooled to a temporary file.
.B mhmail
then invokes
by default, with the name of the temporary file as its argument, to
by default, with the name of the temporary file as its argument to
deliver the message to the specified address.
.PP
The
.B \-subject
switch can be used to specify the \*(lqSubject:\*(rq field of the message.
\*(lqSubject:\*(rq field of the message.
.PP
The
.B \-resent
switch indicates that the message body is a fully formed message that
will be resent, as
.B dist
would do.  For example:
.PP
.RS 5
.nf
mhmail u2@example.com -resent <\`mhpath cur\`
.fi
.RE
.PP
By default,
.B mhmail
will read the message to be sent from the standard input.
You can specify the text of the message at the command
line with the
.B \-body
.I text
switch.  If the standard input has zero length,
.B mhmail
will not send the message and return with status 1.  You can use the switch
.B \-body
\*(lq\*(rq to force an empty message.
.PP
Normally, addresses appearing as arguments are put in the \*(lqTo:\*(rq
field.  At least one such address is required when sending a message,
i.e., if any other switches are supplied.  If the
.B \-cc
switch is used, all addresses following it, even if there are
intervening switches other than
.BR \-to ,
are placed in the \*(lqcc:\*(rq field.  The optional
.B \-to
switch can appear in front of any addresses to signify their placement
in the \*(lqTo:\*(rq field.  The
.B \-to
and
.B \-cc
switches may be given multiple times.
.PP
By using
.B \-from
.IR addr ,
you can specify the \*(lqFrom:\*(rq header of the draft.  If no
.B \-from
switch is used when sending,
.B mhmail
will supply a \*(lqFrom:\*(rq header field using the sender's local mailbox,
see
.I localmbox
in
.IR mh-format (5).
Naturally,
.B post
will fill in the \*(lqSender:\*(rq header correctly.
.PP
The
.B \-headerfield
.I name:value
switch adds a header field with the specified name and value to the
message.  (The
.I value
is called the
.I field body
in RFC 2822, but that's too easily confused with the message body.)
There need not be a space after the colon that separates the
name and value in the argument to this switch.  It is usually best to
enclose the argument in quotes to protect it from the shell.  The
.B \-headerfield
switch may be given multiple times to add multiple header fields,
for example,
.PP
.RS 5
.nf
mhmail -from sender@example.com \\
-headerfield 'MIME-Version:1.0' \\
-headerfield 'Content-Type:text/plain; charset=utf-8' \\
-headerfield 'Content-Transfer-Encoding:8bit' \\
recipient@example.com
.fi
.RE
.PP
Note about this example:  the
.B \-attach
switch causes
.I MIME-Version
and
.I Content-Type
headers to be added to the message.
.PP
The
.B \-profile
switch instructs
.B mhmail
to use
.B send
instead of
.B post
to send the message.  This allows use of the user's context and aliases.
.PP
The
.B \-attach
.I file
switch attaches the specified file to the message.  Use of
.B \-attach
enables
.B \-profile
because
.B mhmail
uses
.B send
to handle the attachment.  It is usually best to enclose the argument
in quotes to protect it from the shell.  The
.B \-attach
switch may be given multiple times to add multiple attachments.
.PP
The
.B \-nosend
option shows the draft file that would be posted/sent on the command's
standard output but does not post or send it.
.B \-send
provides the default behavior of posting or sending the message.  It
can be used to disable
.BR \-nosend ,
for example, when using a shell alias.
.PP
All other switches are passed to
.B post
if
.B \-profile
is not used, or
.B send
if
.B \-profile
is used.
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^%bindir%/inc~^Program to incorporate mail drop into folder
^%nmhlibexecdir%/post~^Program to deliver a message
^/tmp/mhmail*~^Temporary copy of message
.fi
.SH "PROFILE COMPONENTS"
.B mhmail
does
.I not
consult the user's
.I \&.mh_profile
.SH "SEE ALSO"
.IR dist (1),
.IR send (1),
.IR mh-format (5),
.IR send (1),
.IR post (8)
.SH DEFAULTS
.nf
.RB ` "\-from localmbox" '
.RB ` \-profile "' is enabled by default only with " \-attach
.RB ` \-send '
.fi
.SH CONTEXT
If
.B inc
is invoked, then
.BR inc 's
context changes occur.  With the
.B \-profile
switch, the context of
.B send
is used.
.B mhmail
does not read the context file itself.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

uip/mhmail.c

/*
 * mhmail.c -- simple mail program
 *
 * This code is Copyright (c) 2002, by the authors of nmh.  See the
 * COPYRIGHT file in the root directory of the nmh distribution for
 * complete copyright information.
 */

#include <h/mh.h>
#include <h/mts.h>
#include <h/signals.h>
#include <h/utils.h>
#include <signal.h>

static struct swit switches[] = {
#define	BODYSW             0
    { "body text", 0 },
#define	CCSW               1
    { "cc addrs ...", 0 },
#define	FROMSW             2
    { "from addr", 0 },
#define	SUBJSW             3
    { "subject text", 0 },
#define VERSIONSW          4
    { "version", 0 },
#define	HELPSW             5
    { "help", 0 },
#define	RESNDSW            6
    { "resent", -6 },
#define	QUEUESW            7
    { "queued", -6 },
    { NULL, 0 }
};

static char tmpfil[BUFSIZ];

/*
 * static prototypes
 */
static void intrser (int);


int
main (int argc, char **argv)
{
    pid_t child_id;
    int status, iscc = 0, nvec;
    size_t i;
    int queued = 0, resent = 0, somebody;
    char *cp, *tolist = NULL, *cclist = NULL, *subject = NULL;
    char *from = NULL, *body = NULL, **argp, **arguments;
    char *vec[5], buf[BUFSIZ];
    FILE *out;
    char *tfile = NULL;

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* foil search of user profile/context */
    if (context_foil (NULL) == -1)
	done (1);

    /* If no arguments, just incorporate new mail */
    if (argc == 1) {
	execlp (incproc, r1bindex (incproc, '/'), NULL);
	adios (incproc, "unable to exec");
    }

    arguments = getarguments (invo_name, argc, argv, 0);
    argp = arguments;

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [addrs ... [switches]]",
			invo_name);
		    print_help (buf, switches, 0);
		    done (0);
		case VERSIONSW:
		    print_version(invo_name);
		    done (0);

		case FROMSW: 
		    if (!(from = *argp++) || *from == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case BODYSW: 
		    if (!(body = *argp++) || *body == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case CCSW: 
		    iscc++;
		    continue;

		case SUBJSW: 
		    if (!(subject = *argp++) || *subject == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case RESNDSW: 
		    resent++;
		    continue;

		case QUEUESW: 
		    queued++;
		    continue;
	    }
	}
	if (iscc)
	    cclist = cclist ? add (cp, add (", ", cclist)) : getcpy (cp);
	else
	    tolist = tolist ? add (cp, add (", ", tolist)) : getcpy (cp);
    }

    if (tolist == NULL)
	adios (NULL, "usage: %s addrs ... [switches]", invo_name);

    tfile = m_mktemp2(NULL, invo_name, NULL, &out);
    if (tfile == NULL) adios("mhmail", "unable to create temporary file");
    chmod(tfile, 0600);
    strncpy (tmpfil, tfile, sizeof(tmpfil));

    SIGNAL2 (SIGINT, intrser);

    fprintf (out, "%sTo: %s\n", resent ? "Resent-" : "", tolist);
    if (cclist)
	fprintf (out, "%scc: %s\n", resent ? "Resent-" : "", cclist);
    if (subject)
	fprintf (out, "%sSubject: %s\n", resent ? "Resent-" : "", subject);
    if (from)
	fprintf (out, "%sFrom: %s\n", resent ? "Resent-" : "", from);
    else
	fprintf (out, "From: %s\n", getlocalmbox ());
    if (!resent)
	fputs ("\n", out);

    if (body) {
	fprintf (out, "%s", body);
	if (*body && *(body + strlen (body) - 1) != '\n')
	    fputs ("\n", out);
    } else {
	for (somebody = 0;
		(i = fread (buf, sizeof(*buf), sizeof(buf), stdin)) > 0;
		somebody++)
	    if (fwrite (buf, sizeof(*buf), i, out) != i)
		adios (tmpfil, "error writing");
	if (!somebody) {
	    unlink (tmpfil);
	    done (1);
	}
    }
    fclose (out);

    nvec = 0;
    vec[nvec++] = r1bindex (postproc, '/');
    vec[nvec++] = tmpfil;
    if (resent)
	vec[nvec++] = "-dist";
    if (queued)
	vec[nvec++] = "-queued";
    vec[nvec] = NULL;

    for (i = 0; (child_id = fork()) == NOTOK && i < 5; i++)
	sleep (5);

    if (child_id == NOTOK) {
	/* report failure and then send it */
	adios (NULL, "unable to fork");
    } else if (child_id) {
	/* parent process */
	if ((status = pidXwait(child_id, postproc))) {
	    fprintf (stderr, "Letter saved in dead.letter\n");
	    execl ("/bin/mv", "mv", tmpfil, "dead.letter", NULL);
	    execl ("/usr/bin/mv", "mv", tmpfil, "dead.letter", NULL);
	    perror ("mv");
	    _exit (-1);
	}
	unlink (tmpfil);
	done (status ? 1 : 0);
    } else {
	/* child process */
	execvp (postproc, vec);
	fprintf (stderr, "unable to exec ");
	perror (postproc);
	_exit (-1);
    }

    return 0;  /* dead code to satisfy the compiler */
}


static void
intrser (int i)
{
    unlink (tmpfil);
    done (i != 0 ? 1 : 0);
}

